x-pack/plugins/security/server/lib/authentication/providers/kerberos.ts

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Boom from 'boom';
+import { Legacy } from 'kibana';
+import { get } from 'lodash';
+import { canRedirectRequest } from '../../can_redirect_request';
+import { getErrorStatusCode } from '../../errors';
+import { AuthenticationResult } from '../authentication_result';
+import { DeauthenticationResult } from '../deauthentication_result';
+import { BaseAuthenticationProvider } from './base';
+
+/**
+ * The state supported by the provider.
+ */
+interface ProviderState {
+  /**
+   * Access token users get in exchange for SPNEGO token and that should be provided with every
+   * request to Elasticsearch on behalf of the authenticated user. This token will eventually expire.
+   */
+  accessToken: string;
+}
+
+/**
+ * If request with access token fails with `401 Unauthorized` then this token is no
+ * longer valid and we should try to refresh it. Another use case that we should
+ * temporarily support (until elastic/elasticsearch#38866 is fixed) is when token
+ * document has been removed and ES responds with `500 Internal Server Error`.
+ * @param err Error returned from Elasticsearch.
+ */
+function isAccessTokenExpiredError(err?: any) {
+  const errorStatusCode = getErrorStatusCode(err);
+  return (
+    errorStatusCode === 401 ||
+    (errorStatusCode === 500 &&
+      err &&
+      err.body &&
+      err.body.error &&
+      err.body.error.reason === 'token document is missing and must be present')
+  );
+}
+
+/**
+ * Parses request's `Authorization` HTTP header if present and extracts authentication scheme.
+ * @param request Request instance to extract authentication scheme for.
+ */
+function getRequestAuthenticationScheme(request: Legacy.Request) {
```

Comment:
__note:__ experimenting with another way of handling `headerNotRecognized` we have in token and saml providers.

Reply:
The way you have it right now is definitely easier for me to comprehend, I like it!